File: javascript-solutions/leetcode/binarySearch.js

// find index of n in arr. return -1 if n is not found
function index_of(arr, n) {
    let l = 0;
    let r = arr.length - 1;

    while (l <= r) {
        let m = Math.floor((l + r) / 2);
        if (arr[m] < n) {
            l = m + 1;
        } else if (arr[m] > n) {
            r = m - 1;
        } else {
            return m;
        }
    }
    return -1;
}

const arr = [-5,-4, -3, 1, 3, 4, 5];
console.log("index of 3 is", index_of(arr, 3));
console.log("index of 10 is", index_of(arr, 10));

File: rust-solutions/sorts/src/bin/binary-search.rs
use std::cmp::Ordering;

/// basic binary search on a sorted array

fn index_of(arr: &[i32], n: i32) -> Option<usize> {
    let mut l = 0;
    let mut r: usize = arr.len() - 1;

    while l <= r {
        let m = (l + r) / 2;
        match arr[m].cmp(&n) {
            Ordering::Less => l = m + 1,
            Ordering::Greater => r = m - 1,
            _ => return Some(m)
        }
    }
    None
}

fn main() {
    let arr = [-5,-4, -3, 1, 3, 4, 5];
    println!("index of 1 is {:?}", index_of(&arr, 1));
}